galley/pkg/config/analysis/diag/message.go

```diff
@@ -76,6 +77,17 @@ func (m *Message) String() string {
 		"%v [%v]%s %s", m.Type.Level(), m.Type.Code(), origin, fmt.Sprintf(m.Type.Template(), m.Parameters...))
 }
 
+// MarshalJSON satisfies the Marshaler interface
+func (m *Message) MarshalJSON() ([]byte, error) {
+	result := make(map[string]interface{})
+	result["level"] = m.Type.Level().String()
+	result["code"] = m.Type.Code()
+	result["message"] = fmt.Sprintf(m.Type.Template(), m.Parameters...)
+	result["origin"] = m.Origin
```

Comment:
1. this should probably be `m.Origin.FriendlyName()`
2. You should handle the case where `m.Origin` is nil.
3. ... better yet, just have this function call `m.Unstructured(true)` to get the map, so we don't need to duplicate the logic at all.

Reply:
1. I don't think I agree--if this is intended to be machine readable then I think we should expose the entire origin struct.
2. json.Marshal handles nil pointers gracefully already
3. I think we actually want a different format than Unstructured. In practice we're going to have to standardize on a different format for what's in a message anyways, so like @ayj mentioned hopefully once the struct is cleaned up we can just use the json annotations and get rid of this function.

Reply:
What does the output actually look like when you run this? 

Origin is an interface, not directly a struct, and the only "field" we should be exposing is FriendlyName.

Reply:
Ah! I didn't realize it was an interface. I'm less sure of my opinion now, but I still think that after we tighten up the API that we're exposing it's a non issue.

The output looks something like this:

> ~/istio/istio/istioctl/cmd$ go run istioctl/main.go x analyze testdata/ingress/myservice-ingress.yaml -o yaml
> - code: IST0101
>   level: Error
>   message: 'Referenced gateway not found: "istio-autogenerated-k8s-ingress"'
>   origin:
>     Collection: {}
>     Kind: Ingress
>     Name: {}
>     Version: ""
>   parameters:
>   - gateway
>   - istio-autogenerated-k8s-ingress
> 
> Error: Analyzer found issues.
> exit status 79

Reply:
I think we can trim up some of the unnecessary stuff here. I'm not even sure that "origin" is a good user-facing name, we should probably just use something like "resource" or "affectedResource" and just print the FriendlyName. I also don't think there's any need to print parameters, that's an internal detail of how we generate messages.

Reply:
I think for this PR I'll take your suggestion and just use Unstructured(true) because I want the formatting to land, but think about this use case: I want to write some validation that will run in a ci/cd pipeline that will reject diffs if a certain IST error pops up unless the name of the resource includes "test". It may be possible to just regex the entire message, but if I have the schema of the particular IST message (which I do, it's in messages.yaml) then I know the order of the parameters and I can just grab it directly before it gets interpolated into the message.

Allowing third parties to write robust applications by using the structured data is an important use case, and I think exposing as much as possible can only help. I agree in principle that we should be conservative with the API but in this case I think parameters is probably pretty useful to expose.

Reply:
I'm glad you're thinking so deeply about these CI/CD cases. :) I agree it's important to get this right, and I appreciate your insights. 

I think the bit I'm balking at is defaulting our user-facing structure to match our internal structure. My concern is that exposing internal struct stuff never meant for user consumption is risking a future headache when we decide to change it. For this kind of strongly structured message, it's much easier to add new fields (which will generally be ignored by default) than to get rid of old ones.